lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -91,78 +124,117 @@ TensorInfo<IndexType>::TensorInfo(THCState* state,
   // single-point tensor that we still have to operate on. Reduce to a
   // single point.
   if (firstNonOneDim == -1) {
+    assert(excludeDim == -1);
+
     dims = 1;
     sizes[0] = 1;
     strides[0] = 1;
-    return;
+
+    // Everything effectively got collapsed into this dimension
+    return 0;
   }
 
   // Skip the leading size 1 dims
-  numCollapsed += origDims - 1 - firstNonOneDim;
+  numCollapsed += dims - 1 - firstNonOneDim;
 
   // Now, to determine the other collapsible dims. These are the size/strides
   // of the previous inner non-collapsible dim we encounter.
-  long sizeInner = THCudaTensor_size(state, t, firstNonOneDim);
-  long strideInner = THCudaTensor_stride(state, t, firstNonOneDim);
+  long sizeInner = sizes[firstNonOneDim];
+  long strideInner = strides[firstNonOneDim];
 
   for (int i = firstNonOneDim - 1; i >= 0; --i) {
-    long sizeOuter = (i == reduceDim) ? 1 : THCudaTensor_size(state, t, i);
-    long strideOuter = THCudaTensor_stride(state, t, i);
-
-    // The next outermost dimension can be skipped if size 1
-    if (sizeOuter == 1) {
-      ++numCollapsed;
-      continue;
-    }
+    long sizeOuter = sizes[i];
```

Comment:
Can you add a comment saying that this loop only determines how many dimensions can be collapsed, but it doesn't collapse anything? Similarly, a comment on the loop below saying that it collapses the dimensions would be helpful.